src/data/list/palindrome.lean

```diff
@@ -0,0 +1,65 @@
+/-
+Copyright (c) 2020 Google LLC. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Wong
+-/
+
+import data.list.basic
+open list
+
+/-!
+# Palindromes
+
+This module defines *palindromes*, lists which are equal to their reverse.
+
+The main result is the `palindrome` inductive type, and its associated `palindrome.rec_on` induction
+principle. Also provided are conversions to and from other equivalent definitions.
+
+## References
+
+* [Pierre Castéran, *On palindromes*][casteran]
+
+[casteran]: https://www.labri.fr/perso/casteran/CoqArt/inductive-prop-chap/palindrome.html
+
+## Tags
+
+palindrome, reverse, induction
+-/
+
+variables {α : Type*}
+
+/--
+`palindrome l` asserts that `l` is a palindrome. This is defined inductively:
+
+* The empty list is a palindrome;
+* A list with one element is a palindrome;
+* Adding the same element to both ends of a palindrome results in a bigger palindrome.
+-/
+inductive palindrome : list α → Prop
+| nil : palindrome []
+| singleton : ∀ x, palindrome [x]
+| cons_concat : ∀ x {l}, palindrome l → palindrome (x :: (l ++ [x]))
```

Comment:
```suggestion
| cons_concat : ∀ x {l}, palindrome l → palindrome (x :: l ++ [x])
```
I think the brackets aren't necessary here, both cases are equal by definition.

Reply:
I'm not sure about this. We have a strong policy to use `<` over `>`, even though they're definitionally equal; I'd think that the same reasoning applies here. Can we get a second opinion?

Reply:
Fair enough. For anyone else reading this, `x :: l ++ [x] = (x :: l) ++ [x]`. Also to consider is 
```lean
@[simp] lemma cons_append (x : α) (s t : list α) : (x::s) ++ t = x::(s ++ t) := rfl
```
So I think this should be in the form which interacts with `simp` the best. I'm not sure which one that'd be.

Reply:
Yeah, I've been preferring `x :: (l ++ [y])` as it's the simp normal form. Same as in #3448.